addons/mail/static/src/new/composer/composer_hook.js

```diff
@@ -0,0 +1,161 @@
+/* @odoo-module */
+
+import { useMessaging } from "../messaging_hook";
+import { useComponent, useEffect, useState } from "@odoo/owl";
+
+export function useSuggestion() {
+    const messaging = useMessaging();
+    const comp = useComponent();
+    const self = {
+        clearRawMentions() {
+            self.rawMentions.partnerIds.length = 0;
+        },
+        clearSearch() {
+            Object.assign(self.search, {
+                delimiter: undefined,
+                position: undefined,
+                term: undefined,
+            });
+            self.state.items.length = [];
+        },
+        detect() {
+            const selectionEnd = comp.props.composer.selection.end;
+            const selectionStart = comp.props.composer.selection.start;
+            const content = comp.props.composer.textInputContent;
+            if (selectionStart !== selectionEnd) {
+                // avoid interfering with multi-char selection
+                self.clearSearch();
+            }
+            const candidatePositions = [];
+            // keep the current delimiter if it is still valid
+            if (self.search.position !== undefined && self.search.position < selectionStart) {
+                candidatePositions.push(self.search.position);
+            }
+            // consider the char before the current cursor position if the
+            // current delimiter is no longer valid (or if there is none)
+            if (selectionStart > 0) {
+                candidatePositions.push(selectionStart - 1);
+            }
+            const suggestionDelimiters = ["@", ":", "#", "/"];
+            for (const candidatePosition of candidatePositions) {
+                if (candidatePosition < 0 || candidatePosition >= content.length) {
+                    continue;
+                }
+                const candidateChar = content[candidatePosition];
+                if (candidateChar === "/" && candidatePosition !== 0) {
+                    continue;
+                }
+                if (!suggestionDelimiters.includes(candidateChar)) {
+                    continue;
+                }
+                const charBeforeCandidate = content[candidatePosition - 1];
+                if (charBeforeCandidate && !/\s/.test(charBeforeCandidate)) {
+                    continue;
+                }
+                Object.assign(self.search, {
+                    delimiter: candidateChar,
+                    position: candidatePosition,
+                    term: content.substring(candidatePosition + 1, selectionStart),
+                });
+                self.state.count++;
+                return;
+            }
+            self.clearSearch();
+        },
+        fetch: {
+            inProgress: false,
+            rpcFunction: undefined,
+        },
+        insert(option) {
```

Comment:
Can be clarified with `apply` or `insertInto`, otherwise, it may lead to misunderstanding with inserting something into the suggestion list?

Reply:
It was called `insertSuggestion`, so considering it's the `useSuggestion` hook, looks alright as `insert`.

- `apply` is too generic
- `insertInto` doesn't work because the parameter is not a target

I get that name `insert` might be improved, but I tried to keep same semantics as before, with more concise naming for improved readability.

Reply:
In this case, the way we call the method will be `this.suggestions.insert(option)`, which is kind of confusing to me.

Reply:
maybe it makes more sense to call it `selectAndInsert()`? then we have `this.suggestions.selectAndInsert(option)`

Reply:
`selectAndInsert` doesn't add much from plain `insert`.
I would say that `select` also need a target (select _something_, select by itself is quite a loose concept).

Again, `insert` by itself is not that better, it's implicit that it's based on internal state of `useSelection` (which suggestion is being selected).
`insert` is just the best I see for immediate change in this PR. It's as specific as your suggestions, and it has the advantage to be the most concise, so best choice so far.